### PR TITLE
fs cache: allow to have strict lower boundary for file segment size and download remaining data in the background

### PR DESCRIPTION
**Pull request by kssenii**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Allow to have strict lower boundary for file segment size by downloading remaining data in the background. Minimum size of file segment (if actual file size is bigger) is configured as cache configuration setting `boundary_alignment`, by default `4Mi`. Number of background threads are configured as cache configuration setting `background_download_threads`, by default `2`. Also `max_file_segment_size` was increased from `8Mi` to `32Mi` in this PR.

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
This is an automated comment for commit 96fb7f04cbc94c625229afd2958cea056c563920 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/51000/96fb7f04cbc94c625229afd2958cea056c563920/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Interpreters/Cache/Metadata.cpp [hunk 1]
```diff
@@ -293,6 +303,162 @@ void CacheMetadata::doCleanup()
...
+        return;
+
+    LOG_TEST(
+        log, "Downloading {} bytes for file segment {}",
+        file_segment.range().size() - file_segment.getDownloadedSize(false), file_segment.getInfoForLog());
+
+    auto reader = file_segment.getRemoteFileReader();
+
+    /// If remote_fs_read_method == 'threadpool',
+    /// reader iteself bever owns/allocates the buffer.
```
Comment (nickitat) [inline src/Interpreters/Cache/Metadata.cpp:409]:
```suggestion
    /// reader itself never owns/allocates the buffer.
```

File: src/Interpreters/Cache/Metadata.cpp [hunk 2]
```diff
@@ -293,6 +303,162 @@ void CacheMetadata::doCleanup()
...
+    std::optional<Memory<>> memory;
+    if (reader->internalBuffer().empty())
+    {
+        memory.emplace(DBMS_DEFAULT_BUFFER_SIZE);
```
Comment (nickitat) [inline src/Interpreters/Cache/Metadata.cpp:413]: not very important, but would be better to allocate buffer once for a thread.

File: src/Interpreters/Cache/Metadata.cpp [hunk 3]
```diff
@@ -293,6 +303,162 @@ void CacheMetadata::doCleanup()
...
+        {
+            int code = e.getErrno();
+            if (code == /* No space left on device */28 || code == /* Quota exceeded */122)
+            {
+                LOG_INFO(log, "Insert into cache is skipped due to insufficient disk space. ({})", e.displayText());
+                return;
+            }
+            throw;
+        }
+    }
```
Comment (nickitat) [inline src/Interpreters/Cache/Metadata.cpp:451]: shouldn't we update segment state?

File: src/Interpreters/Cache/FileCache.cpp [hunk 4]
```diff
@@ -633,27 +643,14 @@ bool FileCache::tryReserve(FileSegment & file_segment, const size_t size)
...
-                it->second.add(segment_metadata);
-
-                freeable_space += segment_metadata->size();
-                freeable_count += 1;
-
-                return PriorityIterationResult::CONTINUE;
-            }
-
-            ProfileEvents::increment(ProfileEvents::FilesystemCacheEvictedFileSegments);
-            ProfileEvents::increment(ProfileEvents::FilesystemCacheEvictedBytes, segment->getDownloadedSize(false));
```
[the remark this answers is not in the document]
Comment (kssenii) [reply src/Interpreters/Cache/FileCache.cpp:653]: they will - in line `FileCache::733`.

File: src/Interpreters/Cache/FileCache_fwd.h [hunk 5]
```diff
@@ -4,7 +4,9 @@
 namespace DB
 {
 
-static constexpr int FILECACHE_DEFAULT_MAX_FILE_SEGMENT_SIZE = 8 * 1024 * 1024;
+static constexpr int FILECACHE_DEFAULT_MAX_FILE_SEGMENT_SIZE = 32 * 1024 * 1024; /// 32Mi
+static constexpr int FILECACHE_DEFAULT_MIN_FILE_SEGMENT_SIZE = 4 * 1024 * 1024; /// 4Mi
```
[the remark this answers is not in the document]
Comment (kssenii) [reply src/Interpreters/Cache/FileCache_fwd.h:8]: yeah, agree, what could be a better name? `FILECACHE_DEFAULT_FILE_SEGMENT_ALIGNMENT`?

File: src/Interpreters/Cache/Metadata.cpp (hunk 3 above)
Comment (kssenii) [reply src/Interpreters/Cache/Metadata.cpp:451]: it is done in `FileSegment::write` method, there is a `try-catch` as well, which updates the file segment state and throws: https://github.com/ClickHouse/ClickHouse/blob/fcadb851c83ea35dd6ecae208c27dc3dc4fecf0b/src/Interpreters/Cache/FileSegment.cpp#L397-L398.
https://github.com/ClickHouse/ClickHouse/blob/fcadb851c83ea35dd6ecae208c27dc3dc4fecf0b/src/Interpreters/Cache/FileSegment.cpp#L533-L539
and if `FileSegment::reserve` fails - file segment state is also updates there: 
https://github.com/ClickHouse/ClickHouse/blob/fcadb851c83ea35dd6ecae208c27dc3dc4fecf0b/src/Interpreters/Cache/FileSegment.cpp#L508-L512

**Review by nickitat — APPROVED** [review]

**Comment by kssenii** [comment]
> Stress test (ubsan) — Killed by signal (in clickhouse-server.log) 

`Logical error: 'Data part 33_27_27_0 is Outdated and has creation TID (85, 33, 08534acf-ca4e-4b19-bc8f-8f9d357d5f0c) and CSN 86, but does not have removal tid. It's a bug or a result of manual intervention.'.`
Issue: https://github.com/ClickHouse/ClickHouse/issues/49923

**Pull request merged**